### ISSUE TITLE
Segmentation fault when *NOT* class = 'DDL'

### ISSUE THREAD

**Issue by nagaura-ryohei** [pgaudit/pgaudit#243]
# Background

We (Fujitsu) sell a DB product (called FEP) that includes PostgreSQL, pgaudit and so on.
We found a bug in FEP. While FEP's PostgreSQL/pgaudit are forked from OSS and modified independently, I think this may be related to OSS pgaudit, so I'll file an issue.
Please bear with us as this is a lengthy explanation.

# Phenomenon
When a database administrator executed "DROP SCHEMA foo_schema CASCADE", a segmentation fault occurred, causing the DB server process to terminate abnormally.
Due to the CASCADE clause, several thousand objects were dropped in a chain reaction.
The user's configuration was "class = 'CONNECT, ERROR'". 
Please note that the SIGSEGV occurred even though DDL was not targeted.
Ideally, the DROP SCHEMA foo_schema CASCADE command should have completed successfully without any audit logging.

The backtrace of the core file is as follows:
```
#0  0x000000000094e29d in pg_strcasecmp ()
#1  0x00007f5c0af490f7 in objecttype_to_bitmap () from ./opt/fsepv14server64/lib/pgaudit.so
#2  0x00007f5c0af4bc01 in apply_all_rules () from ./opt/fsepv14server64/lib/pgaudit.so
#3  0x00007f5c0af46385 in log_audit_event () from ./opt/fsepv14server64/lib/pgaudit.so
#4  0x00000000007f55af in PortalRunUtility ()
#5  0x00000000007f56e3 in PortalRunMulti ()
#6  0x00000000007f5bdf in PortalRun ()
#7  0x00000000007f1cbf in exec_simple_query ()
#8  0x00000000007f363b in PostgresMain ()
#9  0x000000000077218c in PostmasterMain ()
#10 0x00000000004fc518 in main ()
```
pg_strcasecmp() is a PostgreSQL function with the following interface:
```
int
pg_strcasecmp(const char *s1, const char *s2)
```

The core file indicates that SIGSEGV occurred at the point where the first argument s1 was dereferenced. 
This first argument is the same as the first argument of objecttype_to_bitmap().
Tracing back further, it seems that the objectType field of the auditEvent in the top of the auditEventStack was pointing to an invalid area.

# Investigation
We were unable to reproduce this issue in our vendor environment.
We investigated source code and the core file from the user's environment, and we think that the problem may lie in pgaudit_ddl_command_end()/pgaudit_sql_drop() in pgaudit.c.

pgaudit_ddl_command_end() uses SPI to store information at the top of the auditEventStack and performs audit logging using log_audit_event().
However, it creates (and switches to)/(switches to and) deletes a memory context for preprocessing and postprocessing, respectively, making the information stored at the top of the auditEventStack valid only during this period. (The same applies to pgaudit_sql_drop().)

```c
Datum
pgaudit_ddl_command_end(PG_FUNCTION_ARGS)
{
...
    /* Switch memory context for query */
    contextQuery = AllocSetContextCreate(
                            CurrentMemoryContext,
                            "pgaudit_func_ddl_command_end temporary context",
                            ALLOCSET_DEFAULT_MINSIZE,
                            ALLOCSET_DEFAULT_INITSIZE,
                            ALLOCSET_DEFAULT_MAXSIZE);
    contextOld = MemoryContextSwitchTo(contextQuery);
...
    query = "SELECT UPPER(object_type), object_identity, UPPER(command_tag)\n"
            "  FROM pg_catalog.pg_event_trigger_ddl_commands()";
...
   result = SPI_execute(query, true, 0);
...
    for (row = 0; row < SPI_processed; row++)
    {
        HeapTuple    spiTuple;

        spiTuple = SPI_tuptable->vals[row];

        /* Supply object name and type for audit event */
        auditEventStack->auditEvent.objectType =
            SPI_getvalue(spiTuple, spiTupDesc, 1);
        auditEventStack->auditEvent.objectName =
            remove_doble_quote(SPI_getvalue(spiTuple, spiTupDesc, 2));
        auditEventStack->auditEvent.command =
            GetCommandTagEnum(SPI_getvalue(spiTuple, spiTupDesc, 3));

        auditEventStack->auditEvent.logged = false;

        /*
         * Identify grant/revoke commands - these are the only non-DDL class
         * commands that should be coming through the event triggers.
         */
        if (auditEventStack->auditEvent.command == CMDTAG_GRANT ||
            auditEventStack->auditEvent.command == CMDTAG_REVOKE)
        {
            NodeTag currentCommandTag = auditEventStack->auditEvent.commandTag;

            auditEventStack->auditEvent.commandTag = T_GrantStmt;
            log_audit_event(auditEventStack);

            auditEventStack->auditEvent.commandTag = currentCommandTag;
        }
        else
            log_audit_event(auditEventStack);

        if(auditEventStack->auditEvent.objectName)
            pfree(auditEventStack->auditEvent.objectName);
    }
...
    MemoryContextSwitchTo(contextOld);
    MemoryContextDelete(contextQuery);
```

# Detailed Investigation
Here, I'll explain the flow of higher-level functions and why I think the above program may be a cause.

First, pgaudit hooks PostgreSQL's ProcessUtility().
```c
void
_PG_init(void)
{
...
    next_ProcessUtility_hook = ProcessUtility_hook;
    ProcessUtility_hook = pgaudit_ProcessUtility_hook;
```

pgaudit_ProcessUtility_hook() follows this flow:
 - Create an empty AuditEventStackItem in auditEventStack using stack_push() and store basic information in the auditEvent field.
 - Execute PostgreSQL's standard_ProcessUtility(). (We do not consider the existence of other extensions here.)
 - If there are any unlogged audit log events, log them using log_audit_event().

Here, PostgreSQL's standard_ProcessUtility() executes the sql_drop and ddl_command_end event triggers via ProcessUtilitySlow() when DROP SCHEMA is executed.
These event triggers in pgaudit are pgaudit_sql_drop() and pgaudit_ddl_command_end().

As mentioned in the previous section "Investigation", these functions use SPI to store information in a part of the top of the stack and call log_audit_event().
Note that the memory acquired by SPI is only valid within these functions.

log_audit_event() determines whether auditEventStack is subject to audit logging using apply_all_rules().
If it is not subject to audit logging, it returns early.
If it is subject to audit logging, it constructs the logging string and logs it.
However, for the sake of considering scenarios where DDL is not specified in the class, we will not consider this here.

Because of the early return by apply_all_rules(), log_audit_event() called by pgaudit_sql_drop() or pgaudit_ddl_command_end() does not perform audit logging.
Instead, log_audit_event() at the end of pgaudit_ProcessUtility_hook() is triggered.
However, some of the information at the top of auditEventStack has already been free(3)-ed due to the deletion of the memory context.
This causes the mapping between the PostgreSQL process's virtual memory and physical memory to disappear, resulting in a SIGSEGV when accessing that area.

# Phenomenon revisited

Based on our previous investigation, I believe that the condition for this issue to occur is probabilistic and happens when both of the following criteria are met:
 - 'DDL' is not specified to the 'class' parameter.
 - A DDL is executed.

# Other Considerations
The segmentation fault may be related to high memory consumption, such as when many objects are dropped due to CASCADE.
In other words, the system may be prone to crashing when a large amount of heap memory is freed (3).

# Apology
Unfortunately, I cannot share the core file because it belongs to the user's environment.

Also, we do not know how to force the mapping between physical memory and virtual memory.
(I could not find the desired information even after searching.) Therefore, I'm at a loss for how to test or reproduce the issue.

Do you have any thoughts or comments?

**Comment by nagaura-ryohei** [comment]
I investigated this issue with PG14 and its corresponding pgaudit version, but the same situation was observed even in the latest version.

Also, if the investigation in the initial post was correct, I propose the following fix:
```diff
Datum
pgaudit_ddl_command_end(PG_FUNCTION_ARGS)
{
-     contextQuery = AllocSetContextCreate(
-                             CurrentMemoryContext,
-                             "pgaudit_func_ddl_command_end temporary context",
-                             ALLOCSET_DEFAULT_MINSIZE,
-                             ALLOCSET_DEFAULT_INITSIZE,
-                             ALLOCSET_DEFAULT_MAXSIZE);
-     contextOld = MemoryContextSwitchTo(contextQuery);
+     contextOld = MemoryContextSwitchTo(auditEventStack->contextAudit);
...
     MemoryContextSwitchTo(contextOld);
-     MemoryContextDelete(contextQuery);
...
Datum
pgaudit_sql_drop(PG_FUNCTION_ARGS)
{
...
-     contextQuery = AllocSetContextCreate(
-                             CurrentMemoryContext,
-                             "pgaudit_func_ddl_command_end temporary context",
-                             ALLOCSET_DEFAULT_MINSIZE,
-                             ALLOCSET_DEFAULT_INITSIZE,
-                             ALLOCSET_DEFAULT_MAXSIZE);
-     contextOld = MemoryContextSwitchTo(contextQuery);
+     contextOld = MemoryContextSwitchTo(auditEventStack->contextAudit);
...
     MemoryContextSwitchTo(contextOld);
-     MemoryContextDelete(contextQuery);
```

However, since I am not familiar with the design of this part of this source code, I may be making an inappropriate suggestion.

**Comment by nagaura-ryohei** [comment]
Hi, this chiket was turned out to be my mistake.
On this project, [1] enables us to avoid my phenomenon.

Thus, I'll close this ticket soon.

[1] https://github.com/pgaudit/pgaudit/blame/REL_14_STABLE/pgaudit.c#L1765
